lib/archethic/contracts/interpreter/version1/common_interpreter.ex

```diff
@@ -0,0 +1,168 @@
+
+  # ----------------------------------------------------------------------
+  #                                _ _
+  #   _ __  _ __ _____      ____ _| | | __
+  #  | '_ \| '__/ _ \ \ /\ / / _` | | |/ /
+  #  | |_) | | |  __/\ V  V | (_| | |   <
+  #  | .__/|_|  \___| \_/\_/ \__,_|_|_|\_\
+  #  |_|
+  # ----------------------------------------------------------------------
+  # the atom marker (set by sanitize_code)
+  def prewalk(:atom, acc), do: {:atom, acc}
+
+  # expressions
+  def prewalk(node = {:+, _, _}, acc), do: {node, acc}
+  def prewalk(node = {:-, _, _}, acc), do: {node, acc}
+  def prewalk(node = {:/, _, _}, acc), do: {node, acc}
+  def prewalk(node = {:*, _, _}, acc), do: {node, acc}
+  def prewalk(node = {:>, _, _}, acc), do: {node, acc}
+  def prewalk(node = {:<, _, _}, acc), do: {node, acc}
+  def prewalk(node = {:>=, _, _}, acc), do: {node, acc}
+  def prewalk(node = {:<=, _, _}, acc), do: {node, acc}
+  def prewalk(node = {:|>, _, _}, acc), do: {node, acc}
+  def prewalk(node = {:==, _, _}, acc), do: {node, acc}
+  def prewalk(node = {:!=, _, _}, acc), do: {node, acc}
+  def prewalk(node = {:++, _, _}, acc), do: {node, acc}
+  def prewalk(node = {:!, _, _}, acc), do: {node, acc}
+
+  # blocks
+  def prewalk(node = {:__block__, _, _}, acc), do: {node, acc}
+  def prewalk(node = {:do, _}, acc), do: {node, acc}
+  def prewalk(node = :do, acc), do: {node, acc}
+
+  # literals
+  # it is fine allowing atoms since the users can't create them (this avoid whitelisting functions/modules we use in the prewalk)
+  def prewalk(node, acc) when is_atom(node), do: {node, acc}
+  def prewalk(node, acc) when is_boolean(node), do: {node, acc}
+  def prewalk(node, acc) when is_number(node), do: {node, acc}
+  def prewalk(node, acc) when is_binary(node), do: {node, acc}
+
+  # converts all keywords to maps
+  def prewalk(node, acc) when is_list(node) do
+    if AST.is_keyword_list?(node) do
+      new_node = AST.keyword_to_map(node)
+      {new_node, acc}
+    else
+      {node, acc}
+    end
+  end
+
+  # pairs (used in maps)
+  def prewalk(node = {key, _}, acc) when is_binary(key), do: {node, acc}
+
+  # maps (required because we create maps for each scope in the ActionInterpreter's prewalk)
+  def prewalk(node = {:%{}, _, _}, acc), do: {node, acc}
+
+  # variables
+  def prewalk(node = {{:atom, var_name}, _, nil}, acc) when is_binary(var_name), do: {node, acc}
+  def prewalk(node = {:atom, var_name}, acc) when is_binary(var_name), do: {node, acc}
+
+  # module call
+  def prewalk(node = {{:., _, [{:__aliases__, _, _}, _]}, _, _}, acc), do: {node, acc}
+  def prewalk(node = {:., _, [{:__aliases__, _, _}, _]}, acc), do: {node, acc}
+
+  # whitelisted modules
+  def prewalk(node = {:__aliases__, _, [atom: module_name]}, acc)
+      when module_name in @modules_whitelisted,
+      do: {node, acc}
+
+  # internal modules (Process/Scope/Kernel)
+  def prewalk(node = {:__aliases__, _, [atom]}, acc) when is_atom(atom), do: {node, acc}
```

Comment:
maybe we should restrict the atom here to avoid ast like:

```elixir
{{:., [line: 1],
  [
    {:__aliases__, [line: 1], [:Kernel]},
    :spawn
  ]}, [line: 1],
 [
   {:fn, [line: 1],
    [{:->, [line: 1], [[], 10]}]}
 ]}
```

```elixir
{{:., [line: 1],
  [
    {:__aliases__, [line: 1],
     [:Process]},
    :exit
  ]}, [line: 1],
 [{:pid, [line: 1], nil}]}
```

which could be dangerous

Reply:
At first I thought the same, but since users cannot create atoms, code is much shorter this way. Should I make the change?

Reply:
It's ok indeed